cpp/inc/bond/ext/grpc/win_thread_pool.h

```diff
@@ -0,0 +1,176 @@
+
+        /// @brief Schedules a callback for execution.
+        ///
+        /// @param callback: functor object to be scheduled.
+        template <typename Callback>
+        void operator()(Callback&& callback)
+        {
+            _impl->schedule(std::forward<Callback>(callback));
+        }
+
+    private:
+        enum class pool_type
+        {
+            new_instance,
+            default_instance
+        };
+
+        class impl
+        {
+        public:
+            impl(pool_type type, boost::optional<std::pair<uint32_t, uint32_t>> numThreads)
+                : _pool{ nullptr, CloseThreadpool },
+                  _group{ nullptr, CloseThreadpoolCleanupGroup }
+            {
+                InitializeThreadpoolEnvironment(&_env);
+
+                if (type == pool_type::new_instance)
+                {
+                    _pool.reset(CreateThreadpool(nullptr));
+                    if (!_pool)
+                    {
+                        detail::Win32Exception("Failed to create thread pool.");
+                    }
+
+                    SetThreadpoolCallbackPool(&_env, _pool.get());
+                }
+
+                _group.reset(CreateThreadpoolCleanupGroup());
+                if (!_group)
+                {
+                    detail::Win32Exception("Failed to create thread pool cleanup group.");
+                }
+
+                SetThreadpoolCallbackCleanupGroup(&_env, _group.get(), nullptr);
+
+                if (numThreads)
+                {
+                    if (!SetThreadpoolThreadMinimum(_pool.get(), static_cast<DWORD>(numThreads->first)))
+                    {
+                        detail::Win32Exception("Failed to set thread pool minimum number of threads.");
+                    }
+
+                    SetThreadpoolThreadMaximum(_pool.get(), static_cast<DWORD>(numThreads->second));
+                }
+            }
+
+            ~impl()
+            {
+                CloseThreadpoolCleanupGroupMembers(_group.get(), FALSE, nullptr);
+                DestroyThreadpoolEnvironment(&_env);
+            }
+
+            impl(const impl& other) = delete;
+            impl& operator=(const impl& other) = delete;
+
+            template <typename Callback>
+            void schedule(Callback&& callback)
+            {
+                using Func = typename std::decay<Callback>::type;
+
+                std::unique_ptr<Func> func{ new Func{ std::forward<Callback>(callback) } };
+
+                if (!TrySubmitThreadpoolCallback(
+                        static_cast<PTP_SIMPLE_CALLBACK>([](PTP_CALLBACK_INSTANCE, PVOID context)
+                        {
+                            (*std::unique_ptr<Func>{ static_cast<Func*>(context) })();
+                        }),
+                        func.get(),
+                        &_env))
+                {
+                    detail::Win32Exception("Failed to submit thread pool callback.");
+                }
+
+                func.release();
```

Comment:
<strike>If the callback never runs, does this leak a `Func`?</strike>
To answer my own question: no. In the dtor `CloseThreadpoolCleanupGroupMembers` is called with `fCancelPendingCallbacks: FALSE`, so it will block until work is drained.